src/storage/src/hummock/sstable_store.rs

```diff
@@ -267,49 +273,163 @@ impl SstableStore {
+        for idx in block_index..end_index {
+            if self.block_cache.exists_block(object_id, idx as u64) {
+                if min_hit_index > idx && idx > block_index {
+                    min_hit_index = idx;
+                }
+                hit_count += 1;
+            }
+        }
+
+        if hit_count * 3 >= (end_index - block_index) || min_hit_index * 2 > block_index + end_index
+        {
+            end_index = min_hit_index;
+        }
+        stats.cache_data_prefetch_count += 1;
+        stats.cache_data_prefetch_block_count += (end_index - block_index) as u64;
+        let end_offset = start_offset
+            + sst.meta.block_metas[block_index..end_index]
+                .iter()
+                .map(|meta| meta.len as usize)
+                .sum::<usize>();
+        let data_path = self.get_sst_data_path(object_id);
+        let span: await_tree::Span = format!("Prefetch SST-{}", object_id).into();
+        let store = self.store.clone();
+        let join_handle = tokio::spawn(async move {
+            store
+                .read(&data_path, start_offset..end_offset)
+                .verbose_instrument_await(span)
+                .await
+        });
+        let buf = match join_handle.await {
+            Ok(Ok(data)) => data,
+            Ok(Err(e)) => {
+                tracing::error!(
+                    "prefetch meet error when read {}..{} from sst-{} ({})",
+                    start_offset,
+                    end_offset,
+                    object_id,
+                    sst.meta.estimated_size,
+                );
+                return Err(e.into());
+            }
+            Err(_) => {
+                return Err(HummockError::other("cancel by other thread"));
+            }
+        };
+        let mut offset = 0;
+        let mut blocks = VecDeque::default();
+        for idx in block_index..end_index {
+            let end = offset + sst.meta.block_metas[idx].len as usize;
+            if end > buf.len() {
+                return Err(ObjectError::internal("read unexpected EOF").into());
+            }
+            // copy again to avoid holding a large data in memory.
+            let block = Block::decode_with_copy(
+                buf.slice(offset..end),
+                sst.meta.block_metas[idx].uncompressed_size as usize,
+                true,
+            )?;
+            let holder = if let CachePolicy::Fill(priority) = policy {
+                let cache_priority = if idx == block_index {
+                    priority
+                } else {
+                    CachePriority::Low
+                };
+                self.block_cache
+                    .insert(object_id, idx as u64, Box::new(block), cache_priority)
+            } else {
+                BlockHolder::from_owned_block(Box::new(block))
+            };
+
+            blocks.push_back(holder);
+            offset = end;
+        }
+        Ok(Box::new(PrefetchBlockStream::new(blocks, block_index)))
```

Comment:
Btw, the default `prefetch_buffer_capactiy` doesn't need be be very large with this approch because IMO a small buffer can already achieve good enough performance when the prefetch concurrency is not high, which is mostly the case for streaming CN. My opinion:
- For streaming CN with large memory, I think 200MB prefetch buffer is an okay default number. 
- For streaming CN with small memory, prefetch can benefit little so we can have a even smaller number
- For serving CN, prefetch concurrency can be much higher and prefetch can also be more aggressive so the prefetch buffer can be the remaining memory minus block cache capacity, which also makes sense and provides good reason for user to deploy serving CN if they do care about batch query performance.


More brainstorms for future works (just immature ideas):
- We can do adaptive prefetch instead of prefetch a fixed number of blocks each time
- Backfill behaves more like a batch query than a streaming query so it may need more prefetch buffer. We can probably use ad-hoc instance (like a serving CN with large memory) to do the scan to achieve better prefetch.

Reply:
It is strange if we do not pin these block read by prefetch.
Because when we only read one block, we will pin it, and when we read several blocks, we do not pin them anymore.
If we do not hope to pin these blocks, we shal not pin any block in iterator.